### DISCUSSION TITLE
Optimising Detection

### DISCUSSION THREAD

**Discussion by daernsinstantfortress** [mmcc-xx/WhosAtMyFeeder#14] [General]
Thanks for the updated version, which is working well!

I thought I'd start a thread to discuss how to optimise setup to give the best chance of getting a successful match. I've had a couple of cases where it's mis-classified a bird based on a specific capture from Frigate, but when looking at the video clip (great feature BTW!) you can see that Frigate caught a reasonably ambiguous image with which to capture on.

e.g. This bird, which was identified as a "Downy Woodpecker":
![image](https://github.com/mmcc-xx/WhosAtMyFeeder/assets/7900720/17e47f35-92c2-40be-a1f3-1021974f751c)

...but when you check the video clip, you can see that it's actually just a common garden tit (a great tit, I think!):
https://github.com/mmcc-xx/WhosAtMyFeeder/assets/7900720/c2299fca-e337-47a2-944c-9b0e04083ba4

Perhaps this is just a limitation of Frigate saying "good enough for a bird", rather than picking the best capture of a bird it can for investigation. Anyway, thought it would make for an interesting discussion.

**Comment by mmcc-xx** [comment]
(That's great quality video - what kind of camera are you using?)

Thanks for bringing this up! I've had similar issues and I've been thinking about how it might be addressed. Frigate starts an event and sends out an MQTT message when an object type overcomes its threshold for, say, birdy-ness, and then sends update MQTT messages when the object moves significantly or when the birdy-ness score improves (I think that's how it works anyway). What this app does is grab snapshot, cropped to the bounding boxes of the bird, when it gets these MQTT messages. The ID with the best score get associated with the event. Frigate decides which image to associate with the event on its own. So, the ID might not have been based on the image you see in the UI, but on one of the intermediary images.

This generally seems to work OK, but it could be better. Some thoughts...

* The app could potentially store all IDs associated with the event with scores above a threshold, and rank them in the UI.

* The app could grab the completed event's video clip and analyze every frame, though that would be rather CPU expensive.

* When an event is live, the app could hit this Frigate API in a loop and analyze the returned frames:
GET /api/<camera_name>/latest.jpg[?h=300]
The most recent frame that Frigate has finished processing. It is a full resolution image by default.

However, those images aren't object specific, or event specific, and are not cropped.

* We could ask the Frigate folks kindly for a new API call that would act like the previous one, but would be for a specific event, and would return images cropped to the detected object.

I'm kinda liking a combo of the first option and the last option, but I'm very open to thoughts.

Also, I could save off the image(s) with the best species scores instead of relying on Frigate's snapshots.

If there are other classification models around I'm happy to try them. I haven't found any others. This one works pretty well, but it would be nice if it took location into account - it is calling my "Ruby Throated Hummingbirds" "Anna's Hummingbirds" and they don't live around here :)

**Reply by daernsinstantfortress** [reply]
It would be worth nudging the Frigate guys ( @NickM-27 has been playing with this I see!). As I understand it, there's a bit of a trade-off between rapidity of detection and accuracy, so it can categorise something more than once, sometimes undoing its own previous decision. If you're using it (as I am) to turn on security lights when it sees a "person" but not a "cat", then you need it to be quick (i.e. sub-second), but for your use case it's better if it's actually much slower and more accurate, waiting until it's completed the entire process before triggering across to your app so you get a better quality of detection.

e.g. [from the documentation](https://docs.frigate.video/integrations/mqtt):
> frigate/events
> Message published for each changed event. The first message is published when the tracked object is no longer marked as a false_positive. When Frigate finds a better snapshot of the tracked object or when a zone change occurs, it will publish a message with the same id. When the event ends, a final message is published with end_time set.

Are you using the end of event detection? I can't remember exactly, but I seem to recall that this is more likely to give a more accurate result, but will obviously only trigger after the bird has departed. I'm not sure how much data Frigate puts through the model, but perhaps there is scope here to do more than just single frame analysis, albeit at a cost of probably needing a TPU to do it without nailing the CPU :/

> If there are other classification models around I'm happy to try them. I haven't found any others. This one works pretty well, but it would be nice if it took location into account - it is calling my "Ruby Throated Hummingbirds" "Anna's Hummingbirds" and they don't live around here

Yeah, I'm expecting to see a few eclectic breeds that aren't normally visitors to the UK! Right now though, it's mostly wood pigeons!

> (That's great quality video - what kind of camera are you using?)

Reolink RLC-811a. £95 from Amazon, 4k @ 25fps and quite good quality, although clearly my positioning isn't great for contrast right now. It's screwed to a tree for testing ;-) 
![image](https://github.com/mmcc-xx/WhosAtMyFeeder/assets/7900720/b3ef97b1-7d90-4f53-baf7-ebc83336a8c8)
I'm also pondering one of these for another feeder project: https://www.amazon.co.uk/Green-Feathers-Camera-Mobile-Access/dp/B0779J3FS9/ - I can see me ending up with loads of them around the garden!

**Reply by NickM-27** [reply]
That's not quite correct, by the time the MQTT event is fired the object is considered a true positive by frigate and the label for that particular event will not change. The snapshots are updated at certain points with newer "best" but that is mainly a function of time and some bit of score.

In general, I would suggest an approach of maybe re-running detection multiple times for each object, perhaps on each MQTT update check if the score is higher than the previous image and if so run inference again.

I would not necessarily agree that the `end` will definitely have a better image than the `start` especially since the image where Frigate detects a high score bird won't necessary be the same as the image with the most accurate bird species detection.

**Reply by daernsinstantfortress** [reply]
Haha! Thanks for correcting me - I thought I'd probably get it wrong!

The multiple detection process sounds a good idea. It would be interesting to see how many MQTT updates a typical visit of a bird would generate, but it does feel that the final solution will need to process multiple images in some way to get the bird's best angle for recognition.

**Reply by NickM-27** [reply]
MQTT updates only happen in certain circumstances:
1. new object
2. the object enters a zone
3. the object becomes stationary (and an update every 60 seconds while stationary)
4. the object was stationary and just moved
5. the object is no longer visible

**Reply by mmcc-xx** [reply]
The app is grabbing a snapshot from frigate for a given event for every event update including the end. The best species id score wins.

For this use case, would a frigate api call like /api/event/id/latest.jpg be useful, with arguments for allowing the images to be cropped to the detected object? I was thinking I’d call it every 1/2 second or so while the event was active and maybe check those images as well as the snapshots. 

It seems like the best images for determining if an object is a bird aren’t necessarily the best for determining the species of the bird.

**Reply by NickM-27** [reply]
Yeah, that might be on the higher end of what I'd suggest. It would also be difficult since currently there is no way to get to the second coordinates for an object. 

But what you have now sounds pretty good as is.

**Reply by mmcc-xx** [reply]
> It would also be difficult since currently there is no way to get to the second coordinates for an object.

The GET /api/<camera_name>/latest.jpg[?h=300] api call has an option for providing bounding boxes - I’d have thought what I proposed was sort of a different lens on the same thing. E.g., if someone requests eventxyz/latest.jpg, frigate knows event xyz is from camera abc and pertains to birds, so return the “latest.jpg” image from camea abc cropped to the bird object (or the bird object that’s associated with event xyz).

Am I oversimplifying this?

**Reply by NickM-27** [reply]
Ah yes, it will show current bounding boxes, not sure about cropping though I'll have to check.

**Reply by NickM-27** [reply]
So it can be resized with a different height but it can't be set to crop on a particular part of the image. Now sure how well the identifier would work if it was run on the whole image

and the image would still need to be cropped somewhere since the input to the species model is square I assume

**Reply by mmcc-xx** [reply]
It appears to me that a new API call is needed, but all of the information needed to service the API call is available - if the boxes can be drawn, then a crop can be done, and the appropriate crop needs to be associated with the appropriate event ID if multiple events with different objects are happening at the same time.

I was thinking that I'd take a look at the Frigate API code to see how hard such a thing would be, but... that didn't happen and I'm leaving on a trip in the morning. So, it'll be a couple weeks.

**Reply by NickM-27** [reply]
I'm sorry I was being silly, there already exists an API that would fully satisfy what you want https://docs.frigate.video/integrations/api#get-apieventsidsnapshotjpg

it just may not always be the most up to date, depending on the snapshot best image timeout that is set

**Reply by mmcc-xx** [reply]
That's what I'm using now - my concern is that the snapshot updates when either 1) the object moves significantly or 2) the object detection score improves - or at least that is my understanding.

I'd like to try to get more frames from the event than just the snapshots in case they might provide better species identification even if they aren't better from a bird-as-an-object perspective.

**Comment by daernsinstantfortress** [comment]
So, interestingly I've moved the camera location to get a bit closer to the primary nut feeder (and added a whole lot of protection to keep the squirrels away!), with a view to getting better images and better recognition, but it's actually reduced the number of birds being matched which has left me scratching my head a bit! Frigate is detecting like mad, but relatively few get tagged in Frigate by WhosAtMyFeeder (great feature BTW!) e.g.

![image](https://github.com/mmcc-xx/WhosAtMyFeeder/assets/7900720/d42e4645-5516-4718-9100-ead328e8d775)
(none of these have been tagged by WhosAtMyFeeder)

Sample snapshot of one unmatched capture:
![image](https://github.com/mmcc-xx/WhosAtMyFeeder/assets/7900720/76f3a975-59a9-427d-857d-52c3390ab91b)

Sample video (from the above snapshot):
https://www.youtube.com/watch?v=_AKV7baczGM

I think that the challenge is mostly around getting a good quality snapshot frame that can be passed to the model to get an accurate match and that, all too often, Frigate is not passing the best frame from the detection for this purpose. After all, it's been told to find a bird and it's confident that it's done so, and doesn't care at all if the bird has its back to the camera! I think before when the camera was further away, Frigate was less confident and while the detection rate in Frigate was lower, a higher percentage of these were being matched to a specific bird type. I would estimate that right now I'm only getting 10-20% matches, despite Frigate getting around 95% of the matches correct (i.e. very few false positives).

On a more positive note, when it does match one, the pictures are lovely :-)
![image](https://github.com/mmcc-xx/WhosAtMyFeeder/assets/7900720/b40389a1-111d-4f10-b71e-8d3cf1b6e9fb)

**Comment by mmcc-xx** [comment]
Hi @NickM-27 - I'm finally getting back into thinking about this :)

As @daernsinstantfortress observed, it doesn't looks like the snapshots that Frigate provides are always the best images for identifying species.

Frigate currently has this API:
https://docs.frigate.video/integrations/api/#get-apicamera_namelatestjpgh300

and this API:
https://docs.frigate.video/integrations/api/#get-apieventsidsnapshotjpg

It would be useful to have an API that was kind of a combination of the two. What I have in mind is something like...
`GET /api/events/<id>/latest.jpg`

This would take the same arguments as...
`GET /api/events/<id>/snapshot.jpg`
... but would return the latest frame instead of the latest snapshot. This app could then evaluate many more frames than just those that qualified as improved snapshots.

I imagine that such an API would be useful in other situations where someone is trying to do sub-classification of a detected object.

Based on your knowledge of the Frigate API does this seem feasible?

Thanks!

update: I'm looking at the code for `GET /api/events/<id>/snapshot.jpg` and it looks like it might do exactly what I want it to do... I'm looking into it some more. 
update 2: unfortunately not... it looks like `GET /api/events/<id>/snapshot.jpg` returns what is considered the best frame for the event, not the latest frame, while the event is still in progress.

update 3: One layer deeper than the API definition, I think either the get_current_frame method of the CameraState object needs an option to crop to a specific object type, or CameraState needs a new method for that capability. With that, it would be pretty easy to create an API call to do what I have in mind. I'm happy to give that a try.

**Reply by NickM-27** [reply]
Yeah, it shouldn't be difficult, just need to get the latest frame and crop on that frame

**Reply by mmcc-xx** [reply]
How does Frigate deal with scenarios where multiple instances of the same type of object are found in the same frame? Does it try to track the individual objects and manage multiple events for each individual object, or just keep one event going while at least one instance of the object type is detected? (or maybe something else entirely?)

**Reply by NickM-27** [reply]
`event` is synonymous with tracked object. Every tracked object is a separate event with its own ID and is tracked separately

**Reply by mmcc-xx** [reply]
OK - good to know. I'll probably have more questions when I dig into it.
